### ISSUE TITLE
MonoInstaller Installing Inconsistency 

### ISSUE THREAD

**Issue by Soren025** [ssannandeji/Zenject-2019#96]
Something I noticed about `MonoInstallers` is that when added in the inspector the `CompisitionRoot` simply calls the method on the prefab without instantiating them. On the other hand, if you were to install them in code using `Install<>()`, passing in an instance of `MonoInstaller`. It attempts to load the prefab from resources and instantiates the prefab.

A comment I found at around line 424 of `DIContainer` mentions why

``` csharp
// When running, it is nice to keep the installer around so that you can change the settings live
// But when validating at edit time, we don't want to add the new game object
```

Basically I feel both should do one or the other as both methods of installing a `MonoInstaller` deal with prefabs. Personally I am in preference with the instantiating of the prefabs so editing them in the scene does not persist out of play mode. That and it would be nice to have access to co-routines within a `MonoInstaller` for installing `Facade` containers along side additive scenes using `ZenUtil` methods.

**Comment by Soren025** [comment]
On the flip side, never instantiating and calling on the prefab has advantages too. When calling `Install<>()` on an installer of type `MonoInstaller` for a sub container `Facade`, the resulting instantiated object will persist after the `Facade` has been disposed. Loading the prefab from resources to call without instantiating removes this issue, especially if it is a `FacadeFactory`.

**Comment by svermeulen** [comment]
The way it's intended to be used is for the user to add installer prefabs to their scene and then reference those game objects in the Installers field of the composition root.   I did not intend for users to drag installer prefabs directly on to the composition root.  This seems to work but could result in some unexpected behaviour, since you could be directly changing properties of the prefab in code and have those changes gets saved persistently.

I can detect this when inside the unity editor, so maybe I should throw an exception in this case to prevent users from doing this, and instead require that they either install from resources or add their prefab to the scene and reference that.  What do you think?

That comment refers to the way it deals with dynamically created installers during the validation step.  I think you bring up a good point though - and that it should probably instantiate them during validation to mimic what it does at runtime, then delete it after validation completes.

**Comment by Soren025** [comment]
I was unaware that I was doing it wrong the whole time, It always worked to pass in prefabs to the `SceneCompisitionRoot` so I always did it.

> I can detect this when inside the unity editor, so maybe I should throw an exception in this case to prevent users from doing this, and instead require that they either install from resources or add their prefab to the scene and reference that. What do you think?

I think that would be helpful, It felt more intuitive to add the prefabs directly. I know now not to do it but an error saying you can't would do the job for you.

I must state that the `GlobalCompisitionRoot` does not instantiate the `MonoInstaller` prefabs from the `GlobalInstallerConfig`s and runs the code off the prefab. From what you said should it instantiate those `MonoInstaller`s as well?

I don't quite understand the validation part, but I do believe if the installers are being run on instantiated prefabs, then those instantiated prefabs should not be destroyed. Think about the facade initializer method for `FacadeFactories` or `ToMethod(Action<T>)`. These may call things on the `MonoInstaller` object and may be unwise to destroy them as a result.

**Comment by svermeulen** [comment]
Good point about GlobalCompositonRoot also not instantiating its installers.  I think it probably should.

Directly referencing prefab installers does work, so we could support that, but we may want to disallow it anyway to enforce consistency and to avoid the potential issue of accidentally changing a value at runtime that gets automatically saved to the prefab.  Also, as you say, you can change the settings at runtime without needing to have that change apply persistently.

Yeah, once an installer prefab is instantiated at runtime, it should not be destroyed for the reasons you mention.  I was talking about validation though - which occurs at edit-time not runtime.  The way validation works is that it executes all your installers to create a fully bound container, and then makes sure all the dependencies can properly get resolved, without ever actually instantiating anything.  See [here](https://github.com/modesttree/Zenject#object-graph-validation) for details

So during validation, we could just call prefabs directly or we could instantiate them and then delete them immediately afterwards.

**Comment by Soren025** [comment]
I think I understand the validation part a bit more now. Not enough to know which path to take during validation of `MonoInstaller`s. For some reason I had the impression that validation could occur at run-time as well as edit time. I realize now how silly that sounds.

**Comment by Soren025** [comment]
Only thing I wonder about is how to handle dynamic `MonoInstaller`s being installed onto sub containers. If a new instance of the `MonoInstaller` gets created every call to `Install<SomeMonoInstaller>()`, what happends when the `Facade` initializer method attached to a `FacadeFactory` calls it.

Should the sub container track the dynamically installed `MonoInstaller`s as they accumulate and remove them once the `DIContainer` is unused? Or should the method to dynamically install `MonoInstaller`s first check for and reuse previously instantiated `MonoInstaller`s?

If instantiated `MonoInstaller`s are reused, should they be managed as `DontDestroyOnLoad()`? A root object to hold all dynamic `MonoInstaller`s that persists scenes for reuse?

**Comment by svermeulen** [comment]
Great questions...  I can imagine a scenario where you are creating and destroying a bunch of facades at runtime, and which calls `Container.Install<>` with a `MonoInstaller`.  Those `MonoInstaller` prefabs instantiated from the resources folder would remain in the scene after the facades are destroyed.

I think it would get complex to try and figure out a way to have the newly instantiated `MonoInstaller` be destroyed along with the facade.  Zenject is pretty simple in terms of how involved it is in the lifecycle of objects.  When the `CompositionRoot` object is destroyed, it just calls `Dispose` on all objects bound to `IDisposable` and that's pretty much it.  If there was a game object associated with the `Facade` then we could just parent the `MonoInstaller` to that, but there isn't necessarily one.  The facade could entirely consist of normal C# classes.

So I think you're probably right that we should instantiate each `MonoInstaller` prefab once and then re-use this each time `Container.Install` is called with it.  This would mean that it would get its members injected each time (in case another installer binds a parameter for it) which feels a bit odd but probably is fine.  This would also have the added benefit that you could tweak the settings in this prefab and have the change apply to all facades at once.

We could define an empty game object called 'Installers' underneath the `GlobalCompositionRoot` which can serve as the location for all dynamically created `MonoInstaller`s.  We could also instantiate any `MonoInstaller` prefabs there that are hooked up directly to either the `GlobalCompositionRoot` or `SceneCompositionRoot`.  What do you think?

**Comment by Soren025** [comment]
Sounds good except for one thing that caught my attention. If `MonoInstaller`s placed in a `SceneCompisitionRoot` are suppose to be in the scene already, it wouldn't be necessary to tie those into the the proposed system.

Unless of course you mean to instantiate the `MonoInstaller`s on `SceneCompisitionRoot` if they are physical prefabs. Remember you are unable to check if a `GameObject` is a prefab outside of the editor so this would not work well at runtime.

This is fine I think for global installers as they are required to be prefabs anyway.

**Comment by svermeulen** [comment]
Yes I was assuming I could check whether a game object was a prefab or not outside of the editor.  That's unfortunate.

Yeah, sounds reasonable to just reuse `MonoInstaller`s that are already in the scene

**Comment by svermeulen** [comment]
Actually we may want to instantiate the `MonoInstaller`s from their prefab anyway when they are referenced again, even if they are already in the scene, since some properties may have been changed

**Comment by Soren025** [comment]
In my opinion, scenes and their `Installer`s should remain a closed system. What if a `SceneCompisitionRoot` contains a `MonoInstaller` that exists within the scene but has no prefab?

I would argue that global installers and dynamic installers be separated as well.

Here is my thought process, in no way do I know how viable this is.

| Installer Source | Comments |
| --- | --- |
| `SceneCompisitionRoot` | Require that all `MonoInstaller`s are within the same scene, prefab or not. Force this requirement by checking and throwing an error when adding in the inspector. The reasoning is because I can very well see a `MonoInstaller` existing in the scene that is not bound to a prefab. Also these `MonoInstaller`s will be created with the scene, and die with the scene automatically so no maintenance required. |
| `GlobalInstallerConfig` | All of them are going to be prefabs due being a project asset. Instantiate all of them and put them as children of the `GlobalCompisitionRoot` (possibly in a `GameObject` called "Installers" under the `GlobalCompisitionRoot`) regardless of the existence of dynamic installers of the same type. The reasoning for this is because one can't guarantee that the prefab in the `GlobalInstallerConfig` is the same in "Resources/Installers/" and may have different values (Why someone would do this idk. But I feel allowing this is better than trying to figure whitch one to use over the other). This should only cause a problem when a `MonoInstaller` within the `GlobalInstallerConfig` calls `Install<SomeMonoInstaller>()`  when `SomeMonoInstaller` is also in the `GlobalInstallerConfig`. At that point it will yell at you on startup (May be even in edit-time validation but I don't fully know how that works) |
| `Install<SomeMonoInstaller>()` | Before instantiating, check if cached. If so use the cached version instead of instantiating a new one. Cached `MonoInstaller`s I feel should be kept in an object separate to `GlobalCompisitionRoot` to avoid confusion. Say the object is called "DynamicInstallers". This object would also be set to `DontDestroyOnLoad()`. Basically what all has been mentioned above already. |

What are your thoughts?

**Comment by svermeulen** [comment]
That's a great summary.  I agree - I'll include these changes in Zenject 4

**Comment by Soren025** [comment]
One thing I just remembered, `Install<SomeInstaller>()` accepts extra args to inject. Not only that but I also have forgotten that `Installer` holds a `DIContainer` that gets injected to install onto.

If `MonoInstaller`s are to be cached when dynamic, are there any problems caused by injecting their values (including their `DIContainer`) more than once? If not, how do you think it will effect `MonoInstallers` that need to use those values for lazy injection such as `FacadeFactory` and `ToMethod()` bindings?

**Comment by svermeulen** [comment]
It did occur to me that we would be injecting the dynamic installers twice, which I thought seemed dangerous, but I couldn't think of any bugs with that.  You're right though, there actually are some issues with that.  Anytime injected member variables of the installer are used within a `ToMethod` or `ToFacadeFactory` method, it would only be using the most recently injected values.

**Comment by svermeulen** [comment]
Maybe for dynamic installers, we should just give up on the ability to tweak the settings at run time, and just instantiate it then delete it immediately afterwards

**Comment by svermeulen** [comment]
Maybe for dynamic installers, we should just give up on the ability to tweak the settings at run time, and just instantiate it then delete it immediately afterwards.  Though I'm not sure how this would behave when it comes to late callbacks like those in ToMethod or ToFacadeFactory

Alternatively, we could just spawn multiple copies as necessary and put them under DynamicInstallers like in your original suggestion.  This seems undesirable though because I can imagine scenarios where there are hundreds of facades being created and destroyed.  And every time they are destroyed they would leave an installer game object around that will clutter up the scene

**Comment by Soren025** [comment]
At the very least for `DIContainer`s managed by a `Facade`, the following would solve the issue of installer clutter if the "create new ones every time" route is taken.

``` csharp
public class DynamicMonoInstallerDisposer : IDisposable
{
    private MonoInstaller monoInstaller;

    public DynamicMonoInstallerDisposer(MonoInstaller monoInstaller)
    {
        this.monoInstaller = monoInstaller;
    }

    public void Dispose()
    {
        Destroy(monoInstaller.gameobject);
    }
}

// The following within the `Install<SomeMonoInstaller>()` method
this.Bind<IDisposable>().ToInstance(new DynamicMonoInstallerDisposer(newMonoInstaller));
```

Other than that, perhaps tracking the dynamic `MonoInstaller`s in a list and destroy them all inside `~DIContainer() { }`. Or even create a `DisposeDynamicMonoInstallers()` method if you are in fact managing a `DIContainer` outside the use of a `Facade` (called within `~DIContainer() { }` still).

**Comment by Soren025** [comment]
I feel like this is a "rabbit hole".

**Comment by svermeulen** [comment]
I think this issue has been addressed with the change to make the prefab installers specified separately from the installers in the scene, and also the introduction of GameObjectContext with zenject 4 (please correct me if I'm wrong)
